Add `pad_to_bounding_box` image ops

awsaf49 opened the pull request:
This PR will add `pad_to_bounding_box` functions used in `keras.io/examples` as mention in Issue: #18489. Note, test code hasn't been added yet. Will add the tests once the op gets finalized.

awsaf49 commented:
cc: @fchollet

fchollet commented on keras/ops/image.py:
```diff
@@ -501,3 +501,157 @@ def map_coordinates(
+        image = backend.numpy.expand_dims(image, 0)
+    elif len(image_shape) != 4:
+        raise ValueError(
+            "'image' (shape %s) must have either 3 or 4 dimensions."
```
Use f strings for formatting

awsaf49 replied:
Sorry for the delayed response. I was caught up in a conference deadline. I have updated the code.

fchollet commented on keras/ops/image.py:
```diff
@@ -501,3 +501,157 @@ def map_coordinates(
+
+
+def _pad_to_bounding_box(
+    image, offset_height, offset_width, target_height, target_width, check_dims
```
Does it support tensors with unknown shapes (e.g. with TF)? Can the offsets be scalar tensors?

awsaf49 replied:
With TensorFlow, it does support unknown shapes. I tried this and it worked nicely:

```py
import tensorflow as tf
tf.compat.v1.disable_eager_execution()

image = tf.compat.v1.placeholder(tf.uint8, shape=[None, None, 3])
padded_image = tf.image.pad_to_bounding_box(image, 1, 1, 100, 200)

print(padded_image.shape)  # (100, 200, 3)
```

However, I'm not sure how to add support for all the backends. Any suggestions?

Regarding tensor arguments, `tf.image.pad_to_bounding_box` doesn't have anything specific for tensor arguments, but with tensor `offset`, it works just fine. I tried this and it worked:

```py
import numpy as np
import tensorflow as tf

x = np.arange(12).reshape(2, 2, 3)
one = tf.constant(1)
tf.image.pad_to_bounding_box(x, one, one, 4, 4, True)
```

Similarly, replacing `tf.image.pad_to_bounding_box()` with `keras.ops.image.pad_to_bounding_box()` also works without errors. I also tried it with PyTorch and Jax tensors; `offset` works fine.

fchollet commented on keras/ops/image.py:
```diff
@@ -501,3 +501,157 @@ def map_coordinates(
+        offset_width: Number of columns of zeros to add on the left.
+        target_height: Height of output image.
+        target_width: Width of output image.
+        check_dims: If True, assert that dimensions are non-negative and in
```
We should not need this argument. If shapes are static, we can raise the error in every case (at creation time). If shapes are dynamic (negative dims are not detectable until runtime), let's not assert anything.

awsaf49 replied:
I believe in `tf.image.pad_to_bounding_box`, they determine whether to use static or dynamic dimensions by checking `ndims` in `image.get_shape()` [here](https://github.com/tensorflow/tensorflow/blob/4dacf3f368eb7965e9b5c3bbdd5193986081c3b2/tensorflow/python/ops/image_ops_impl.py#L1116). However, I am unsure how to implement this for multibackend. Any suggestions would be greatly appreciated.

fchollet commented on keras/ops/image.py:
```diff
@@ -501,3 +501,157 @@ def map_coordinates(
+
+
+@keras_export("keras.ops.image.pad_to_bounding_box")
+def pad_to_bounding_box(
```
The API is a bit strange to me -- it says "pad to bounding box" but there's no bounding box being specified.

awsaf49 replied:
Yes I agree, should we consider a different api??

fchollet replied:
I think so -- what could the API look like? `pad_image`?

Also, it is optimal to ask the user to specify `offset_height, offset_width, target_height, target_width`? It could be something else, e.g. `right_padding, left_padding, top_padding, bottom_padding`. Which may be clearer.

awsaf49 replied:
Yes `pad_image` will be nice. I think having `target_width` and `target_height` would be nice in a sense that in a lot of cases we know the final output shape so we won't have to calculate the padding sizes manually.

fchollet replied:
Thanks for the input! Do you think it's better to stay with the TF signature, or expand it with something like

`right_padding=None, left_padding=None, top_padding=None, bottom_padding=None, target_width=None, target_height=None` -- you could specify e.g. `right_padding=..., target_width=...` or `right_padding=..., bottom_padding=...`.

awsaf49 replied:
Apologies for the delay once again; I've just returned home after attending the conference abroad.

I believe expanding the functionality would be a valuable addition for our future work. As a friendly reminder, we need to apply the same changes to the `crop_to_bounding_box` implementation since it uses the same signature:

```py
tf.image.crop_to_bounding_box(
    image, offset_height, offset_width, target_height, target_width
)
```

I will submit a PR for `crop` after this one.

awsaf49 replied:
@fchollet, since `pad_to_bounding_box` and `crop_to_bounding_box` are related functions and we're considering a change in the signature for `pad` from using `offset_height/width` to `left/right/top/bottom_padding`. Should we make a corresponding change for `crop` to use `left/right/top/bottom_cropping` for consistency?

fchollet replied:
Yes, sure. It should likely be called `pad_images` and `crop_images` to distinguish from `ops.pad`.

awsaf49 replied:
Thanks for the feedback, I start adding the features you mentioned.

fchollet reviewed (commented):
Thanks for the PR! It sounds good to add it, it's occasionally used.

awsaf49 commented:
@fchollet I've updated the code to reflect your feedback. Now, it supports `right_padding` and `bottom_padding` along with `target_height` and `target_width`. I've also added the tests. However, the `linting` test fails even though the code is formatted with `black` and `isort`. Weirdly, the linting error is shown for `batch_normalization.py`, which was not affected by any of the commits of this PR.

Error Message:
```shell
Run bash shell/lint.sh
Skipped 27 files
would reformat /home/runner/work/keras/keras/keras/layers/normalization/batch_normalization.py

Oh no! 💥 💔 💥
1 file would be reformatted, 605 files would be left unchanged.
Error: Process completed with exit code 1.
```

fchollet reviewed (commented):
Thanks for the updates!

awsaf49 commented:
@fchollet  I've made the changes to reflect your feedback. Let me know if they are okay.

fchollet commented on keras/ops/image.py:
```diff
@@ -501,3 +501,207 @@ def map_coordinates(
+    if right_padding is None:
+        right_padding = target_width - left_padding - width
+
+    if not top_padding >= 0:
```
Rather than e.g. `if not top_padding >= 0:` you can do `if top_padding < 0:` which is more readable.

fchollet commented on keras/ops/image.py:
```diff
@@ -501,3 +501,207 @@ def map_coordinates(
+
+    batch, height, width, depth = images.shape
+
+    if [top_padding, bottom_padding, target_height].count(None) > 1:
```
Let's switch the `>` to `!=`. This will be more general and it will match the error message perfectly.

fchollet commented on keras/ops/image.py:
```diff
@@ -501,3 +501,207 @@ def map_coordinates(
+    if [top_padding, bottom_padding, target_height].count(None) > 1:
+        raise ValueError(
+            "Must specify exactly two of "
+            "top_padding, bottom_padding, target_height"
```
Also print the values passed:

```
Received: top_padding={top_padding}, bottom_padding={bottom_padding}, target_height={target_height}
```

fchollet commented on keras/ops/image.py:
```diff
@@ -501,3 +501,207 @@ def map_coordinates(
+            "Must specify exactly two of "
+            "top_padding, bottom_padding, target_height"
+        )
+    if [left_padding, right_padding, target_width].count(None) > 1:
```
Same comments here

fchollet commented on keras/ops/image.py:
```diff
@@ -501,3 +501,207 @@ def map_coordinates(
+    elif len(images_shape) != 4:
+        raise ValueError(
+            f"'images' (shape {images_shape}) must have "
+            "either 3 or 4 dimensions."
```
```
Invalid shape for argument `images`: it must have rank 3 or 4. Received: images.shape={images.shape}
```

fchollet commented on keras/ops/image.py:
```diff
@@ -501,3 +501,207 @@ def map_coordinates(
+):
+    images = backend.convert_to_tensor(images)
+    is_batch = True
+    images_shape = images.shape
```
Prefer using `ops.shape`

fchollet commented on keras/ops/image.py:
```diff
@@ -501,3 +501,207 @@ def map_coordinates(
+    if not right_padding >= 0:
+        raise ValueError("width must be <= target - offset")
+    if not bottom_padding >= 0:
+        raise ValueError("height must be <= target - offset")
```
Also print the (invalid) padding values, to help with debugging

awsaf49 commented:
@fchollet thanks for your feedback. I've added the updates you mentioned.

fchollet reviewed (approved):
LGTM, thank you for the great contribution!
